aws/extensions/computetask_awswin_volumemount/extension.ftl

```diff
@@ -0,0 +1,176 @@
+        "_computetask_awswin_volumemount"
+    ]
+    description=[
+        "Uses windows commands to format and mount volumes"
+    ]
+    supportedTypes=[
+        EC2_COMPONENT_TYPE,
+        ECS_COMPONENT_TYPE,
+        COMPUTECLUSTER_COMPONENT_TYPE,
+        BASTION_COMPONENT_TYPE
+    ]
+    scopes=[
+        COMPUTETASK_EXTENSION_SCOPE
+    ]
+/]
+
+[#macro shared_extension_computetask_awswin_volumemount_deployment_computetask occurrence ]
+
+    [#local storageProfile = _context.StorageProfile ]
+    [#local volumes = (storageProfile.Volumes)!{}]
+
+    [#local files = {}]
+    [#local commands = {}]
+
+    [#-- Support for data volumes as linked components to single instances --]
+    [#if occurrence.Core.Type == EC2_COMPONENT_TYPE ]
+
+        [#-- These volumes are provided through DataVolumes which are an indepdent component --]
+        [#-- These are attached to instances through links and VolumeMount macro which applies the extra configuration required --]
+        [#local zoneResources = occurrence.State.Resources.Zones]
+        [#list zones as zone]
+            [#if multiAZ || (zones[0].Id = zone.Id)]
+                [#local zoneEc2InstanceId = zoneResources[zone.Id]["ec2Instance"].Id ]
+                [#list (_context.VolumeMounts)![] as mountId,volumeMount ]
+                    [#local dataVolume = _context.DataVolumes[mountId]!{} ]
+                    [#if dataVolume?has_content ]
+                        [#local zoneVolume = (dataVolume[zone.Id].VolumeId)!"" ]
+                        [#if zoneVolume?has_content ]
+                            [#if ! ( getCLODeploymentUnitAlternative() == "replace1" ) ]
+                                [@createEBSVolumeAttachment
+                                    id=formatDependentResourceId(
+                                        AWS_EC2_EBS_ATTACHMENT_RESOURCE_TYPE,
+                                        zoneEc2InstanceId,
+                                        mountId
+                                    )
+                                    device=volumeMount.DeviceId
+                                    instanceId=zoneEc2InstanceId
+                                    volumeId=zoneVolume
+                                /]
+                            [/#if]
+
+                            [#local volumes += {
+                                mountId : {
+                                    "Enabled" : true,
+                                    "MountPath" : volumeMount.MountPath,
+                                    "Device" : volumeMount.DeviceId,
+                                    "DataVolume" : true
+                                }
+                            }]
+
+                        [/#if]
+                    [/#if]
+                [/#list]
+            [/#if]
+        [/#list]
+    [/#if]
+
+    [#local diskId = -1]
+    [#list volumes as id,volume ]
+
+        [#local deviceId = ""]
+        [#local osMount = ""]
+
+        [#local diskId = diskId + 1]
```

Comment:
Joys of learning a new language - missing the obvious

But a fix in volume mounting, makes this moot - it needs to be restricted to enabled volumes, so there was an introduced bug - now fixed

Reply:
How will this affect enabling and disabling volumes? Would it still be better to use the index value and just have gaps in the Id format?

Reply:
It doesn't, unless I'm misunderstanding how that works. Windows doesn't use the volume names but the incremental number of the volume (based on the order of the enabled volumes, so I have to make sure the numbering aligns is all. eg
```
                "Volumes": {
                    "root" : {
                        "Device" : "/dev/sda1",
                        "Size" : 40
                    },
                    "userdata" : {
                        "Enabled" : false,
                        "Device" : "xvdd",
                        "Size" : 50,
                        "MountPath" : "D"
                    },
                    "data": {
                        "Enabled" : true,
                        "Device": "xvdz",
                        "Size": 10,
                        "MountPath" : "c:\\esdata"
                    }
                }
```
Device /dev/sda1 is Disk0 and device xvdz is Disk1

Reply:
So if thats defined by the Id and I disable and enable a volume on one of my deployments what would happen?

Reply:
Then during that execution (for example) device /dev/sda1 is Disk0, device xvdd would become Disk1 and device xvdz would be removed (assuming you set xvdd to enabled and xvdz to disabled). Disk0 and disk1 are used by diskpart during the userdata run to connect the disks when the EC2 spins up.

Does that answer your question or am I missing something?

Reply:
My question is more how flexible is the disk updates on an existing instance? Can I add and remove disks on an existing instance or will adding a new one force it to mount the disk to another location? 

I.e. if I add a disk in between root and userdata will that disk mount to the location it has been configured to use or will that use the diskId for userdata and end up on the D: drive?

Reply:
It won't force it to another location (since the location is explicit - ie drive letter or path - and separate from the disk number), but there will be an issue with disks as Windows doesn't use raw disks (so they have to be formatted). I'll need to check that existing disks aren't clobbered during an update

Reply:
@ken9ross do we have an answer on your checks on the disks?

Reply:
Testing of this has shown that the instances are replaced when disks are added, so this means that there isn't a risk of disks being remapped and causing data to be overriden. I think we can call this resolved and work through it if it becomes an issue